File: Python/minCost_.py
class Solution:
    def minCost(self, colors: str, neededTime: List[int]) -> int:
        # totalTime: total time needed to make rope coolorful;
        # currMaxTime: maximum time of a balloon needed in this group.
        total_time = 0
        curr_max_time = 0

        # For each balloon in the array:
        for i in range(len(colors)):
            # If this balloon is the first baloon of a new group
            # set the curr_max_time as 0.
            if i > 0 and colors[i] != colors[i - 1]:
                curr_max_time = 0

            # Increment total_time by the smaller one.
            # Update curr_max_time as the larger one.
            total_time += min(curr_max_time, neededTime[i])
            curr_max_time = max(curr_max_time, neededTime[i])

        # return total_time as the minimum removal time.
        return total_time

File: js/minCost.js
/**
 * @param {string} colors
 * @param {number[]} neededTime
 * @return {number}
 */
var minCost = function (s, cost) {
    // Initialize two pointers i, j.
    let totalTime = 0
    let i = 0,
        j = 0

    while (i < cost.length && j < cost.length) {
        let currTotal = 0,
            currMax = 0
        // Find all the balloons having the same color as the
        // balloon index at i, record the total removal time
        // and the maximum removal time.
        while (j < cost.length && s[i] == s[j]) {
            currTotal += cost[j]
            currMax = Math.max(currMax, cost[j])
            j++
        }

        // Once we reach the end of the current group, add the cost of
        // this group to total_time, and reset two pointers.
        totalTime += currTotal - currMax
        i = j
    }
    return totalTime
}
